### ISSUE TITLE
Early Stopping with User Defined Function

### ISSUE THREAD

**Issue by BrianMiner** [aigamedev/scikit-neuralnetwork#95]
Working for a bit now with scikit-neuralnetwork, two things I have enjoyed working with nolearn/Lasagne and I was curious if these are available here and I just cant quite find them are the following...

1) The use of a call to functions whenever an epoch ends - in order to use early stopping criteria via a user defined function or to adjust a learning rate or momentum variable like shown at Daniels blog http://danielnouri.org/notes/category/deep-learning/#changing-learning-rate-and-momentum-over-time.  I find the early stopping feature the most useful 

on_epoch_finished=[
        AdjustVariable('update_learning_rate', start=0.03, stop=0.0001),
        AdjustVariable('update_momentum', start=0.9, stop=0.999),
        EarlyStopping(patience=200),
        ],

2) Printing out performance on the training and test set for each iteration.

Thanks!

**Comment by alexjc** [comment]
1) It's a good idea, I'll look into it after the conference if nobody has done it before then!

2) Isn't possible with PyLearn2, we looked into it and opened ticket:
https://github.com/lisa-lab/pylearn2/issues/1490

It's one of the reasons the new libraries (like keras or deepy) are taking over IMHO, and we've moved to use an interchangeable backend.

**Comment by alexjc** [comment]
For 2) if you don't mind re-running the `score()` function on both training and validation set, that should be easy enough. You can already use `valid_set` and `valid_size` if you want to show some results...

**Comment by BrianMiner** [comment]
I agree with people moving away from pylearn2. nolearn / lasagne can do everything I ever need. Keras, if it can become compatible with scitkitlearn (=especially for grid search / random search) looks very full featured.

**Comment by leconteur** [comment]
Blocks looks promising too. https://blocks.readthedocs.org/en/latest/

**Comment by alexjc** [comment]
@BrianMiner So you think there's good value in wrapping all these different libraries behind a standard interface that's sklearn compatible? The goal is to make it as easy as possible for people to switch, especially if libraries keep changing so fast!

**Comment by BrianMiner** [comment]
Absolutely! It seems to me that nolearn is the most full featured wrapper currently in that it allows a lot of flexibility to access lasagne (the couple things I mentioned above are very useful). It seems to work with scikit grid search (although the print out goes crazy and you need to turn it off or ignore it). Have not tested in scikit pipelines but that is the other key integration with scikit IMHO.

To me seems like Keras has more functionality (e.g. RNN) but not having a convenient way to tune the model inhibits its use. 

I think of an incredible library like xgboost that became much more practical one you could use scikit learn to use a grid or random search (within a pipeline).

**Comment by alexjc** [comment]
Thanks! These `on_epoch` callbacks are not tweakable by sklearn during grid search, are they?

**Comment by BrianMiner** [comment]
To round out my wish list (for what its worth :)) I think the flexibility to train with multi-class outputs (k-hot or a list of one-hots) and access to the weights from the hidden layers (sometimes useful as features).

**Comment by alexjc** [comment]
k-hot is in now :-)

**Comment by BrianMiner** [comment]
If you mean being able to change the functions called by on_epoch I don't think so.

**Comment by alexjc** [comment]
Changed the title since printing validation result is already supported with `valid_set` or `valid_size`, printing training result is not supported by PyLearn2.

**Comment by BrianMiner** [comment]
Since you can access the validation set in the functions that run during on_epoch it is possible (in nolearn at least for lasagne, no idea about Pylearn2) to print out the validation performance for any function you write (e.g. weighted kappa) not just the log loss being used for the model training. Maybe that is a way around the pylearn2 limitation.

**Comment by alexjc** [comment]
You're right, it should just work — though slower. Once user-defined epoch functions are supported, I'll add another ticket if necessary.

**Comment by alexjc** [comment]
Close #133.
